### PR TITLE
NeRF

### PR DESCRIPTION
**Pull request by YanivHollander**
#### Changes

This pull request concerns a draft version of the NeRF algorithm for view synthesis and 3D rendering. 

In this PR I included an API class - CameraCalibration, which drives NeRF, and is designed based on the discussion in #1384. 

I took a brute force approach and **copied** the relevant core algorithms from the paper: Wang et al. (2021) - https://arxiv.org/abs/2102.07064. Most algorithm parameters are hard-coded, whereas some of those can be defined in CameraCalibration API.

I tested the full functionality of CameraCalibration in Colab - https://colab.research.google.com/drive/1xiX1Jf572HAN_TN4p-8IPU3CZRRkfhox#scrollTo=yVT7jsvk4ViS. Please let me know if you can run this notebook, or at least able to copy it and modify the relevant paths to make it work with the NeRF version in this PR.

The next steps in this PR, and I foresee, and please comment on each point:

1. Parameterize the algorithm and export more control to CameraCalibration API (including, e.g., CPU/GPU control, rendering parameters, etc.)

2. Refactor the code I took from https://arxiv.org/abs/2102.07064 to follow Kornia's standards

3. Replace core engine component with of-the-shelve algorithms. Here I mainly refer to the rendering part, which is also the most compute intensive. I'm thinking PyTorch3D may have some good alternatives we may want to explore

Yaniv

#### Type of change

- [ ] 📚  Documentation Update
- [ ] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

**Comment by YanivHollander** [comment]
@edgarriba, I shared the Colab Notebook with you. There is no way I could find in Colab to share with everyone. Should I upload this Notebook to this PR as well?

**Comment by edgarriba** [comment]
> @edgarriba, I shared the Colab Notebook with you. There is no way I could find in Colab to share with everyone. Should I upload this Notebook to this PR as well?

there should be a way to share to anyone with the link. Or yes, add it temporally in this PR

**Comment by edgarriba** [comment]
@YanivHollander would interesting to see (or compare somehow) with classical checkerboard calibration. @ducha-aiki @lferraz do you know any open dataset for that ?

**Comment by YanivHollander** [comment]
The following link should allow anyone to open the Colab:

https://colab.research.google.com/drive/1xiX1Jf572HAN_TN4p-8IPU3CZRRkfhox?usp=sharing

Please let me know if you run into issues opening the Notebook

**Comment by ducha-aiki** [comment]
> @YanivHollander would interesting to see (or compare somehow) with classical checkerboard calibration. @ducha-aiki @lferraz do you know any open dataset for that ?

https://github.com/amy-tabb/basic-camera-calibration here is a code to generate such dataset, thanks to @amy-tabb

Also:
output of the code
https://app.box.com/s/bsg2vm6o3uzqlbbgzfk6w9e27veyel0l
configs used to generate the images above:  https://app.box.com/s/7zbc1098797ubk8mh0q9lg3ual007f3j

**Comment by YanivHollander** [comment]
After more work on the Nerf algorithm, I submitted to this PR few more modifications. Mostly I added more functions to the class, and exposed parameters from the algorithm itself.

I also tried to run the Nerf implementation on one of the calibration datasets proposed by @ducha-aiki. I am still struggling to get satisfying results. For the llff flower dataset, the results are reasonable, but still behind what is presented in the paper. My main problem is lack of compute. I am running the algorithm in Colab CUDA, but this resource is limited. Any suggestions will be greatly appreciated. This is especially important when trying out a deeper Nerf NN over the tiny Nerf currently the default that runs.

Few suggestion points for further discussion:

1.	Per a previous discussion on the Stoke framework, I believe this implementation of Nerf can nicely scale to work in the context of Stoke or a similar framework. I’m mainly referring to the fact the algorithm includes a training process, which could be wrapped by Stoke to automatically tune compute based on device resources

2.	Following another discussion on Nerf shared by @edgarriba on Linkedin, I believe this current contribution can be expanded at some point to include more model implementations, such as GAN based Nerf published in: https://github.com/MQ66/gnerf

But those extensions should be considered only after I improve basic results. Again, I stress that I believe my main problem is compute, so any suggestions in this direction will be greatly appreciated!

**Comment by edgarriba** [comment]
@YanivHollander could you include in this PR the python training script so that I can try in my machine ? possibly a bash or something to easily download the dataset too

**Comment by YanivHollander** [comment]
I shared the location of two datasets I tested:

Link to LLFF flower dataset:
https://drive.google.com/drive/folders/1YflPSvCiInSt4Y-hAmP6ZM4745P58l6r?usp=sharing

Link to one of the calibration datasets:
https://drive.google.com/drive/folders/1skJBBP23_wO5jPuidl6F1cSbVFP_tJax?usp=sharing

The following notebook is shared with everyone:
https://colab.research.google.com/drive/1xiX1Jf572HAN_TN4p-8IPU3CZRRkfhox?usp=sharing

Will this allow you to run some tests on your machine? You should run in the Nerf branch with the code in this PR (that is also committed to YanivHollander:nerf)

**Comment by YanivHollander** [comment]
Hi,

Over the last couple of weeks, I ran several Nerf computations for different scenes. I found a solution for my compute problem with Colab Pro that works quite neatly.

My conclusion at this point is that the current implementation of the Nerf algorithm is not quite one-size-fits-all. There are scenes that are processed well; for example, the flower scene (results attached). On the other hand, I couldn’t quite get good reconstruction results with the calibration data set. Also, I tried to take few 2D shots of a mini superman figure I have, and again with some limited success (results attached).

During my research into how to improve results I added few features to the original algorithm, listed here: 

1.	The ability to save checkpoints during training, and to start training from a previously saved checkpoint
2.	An extension to the focal model that allows each camera to have its own focal lengths that are estimated separately during training
3.	The ability to initialize the Nerf model after training, while keeping focal and pose estimations. This feature allows better training of the model in the second pass (from Wang, 2021 section 4.2).
4.	Few more algorithm parameters are exposed in the API (compute device, number of rays to randomly march from the image plane, learning rate)
5.	A bunch of unit tests

I think that at this point we may want to transition this PR to a regular one and start the process of integrating to Kornia. This will mainly open it up for others to test. I also have few other directions in mind, which I would like to test based on this version (moving to tools from PyTorch3D; implementing prior knowledge with Semantic Consistency Loss - https://www.notion.so/DietNeRF-Putting-NeRF-on-a-Diet-4aeddae95d054f1d91686f02bdb74745; etc.).

Please share your thoughts on how to proceed.

Yaniv

**Comment by YanivHollander** [comment]
llff_flower_img
llff_flower_depth

**Comment by YanivHollander** [comment]
superman_img_superman - 16 png

superman - 16


superman_img_superman - 21 png


superman - 21

File: kornia/geometry/nerf/nerf.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+
+class RayParameters:
+    def __init__(self):
+        self.NEAR, self.FAR = 0.0, 1.0  # ndc near far
+        self.N_SAMPLE = 128  # samples per ray
+        self.POS_ENC_FREQ = 10  # positional encoding freq for location
+        self.DIR_ENC_FREQ = 4  # positional encoding freq for direction
+
+
+class LearnFocals(nn.Module):
```
Comment (edgarriba) [inline kornia/geometry/nerf/nerf.py:26]: wondering how we could leverage here the `PinholeCamera` data structure
https://kornia.readthedocs.io/en/latest/geometry.camera.pinhole.html#kornia.geometry.camera.pinhole.PinholeCamera

File: kornia/geometry/nerf/nerf.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+            fx = self.__fxfy[0][0]
+            fy = self.__fxfy[0][1]
+        fxfy = torch.stack([fx**2 * self.__w, fy**2 * self.__h])
+        return fxfy
+
+    def is_multi_focal(self):
+        return self.__multi_focal
+
+
+class CameraCalibration:
```
Comment (edgarriba) [inline kornia/geometry/nerf/nerf.py:53]: possibly start thinking how to integrate as a trainer
```suggestion
class CameraCalibration(kornia.x.Trainer):
```

File: kornia/geometry/nerf/nerf.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+
+        self._epoch0 = 0
+
+    def load_images(self, path, num_of_images_to_use: int = None):
+        images: List[np.array] = []
+        self._image_filenames: List[str] = []
+        for i, filename in enumerate(os.listdir(path)):
+            if num_of_images_to_use is not None and i >= num_of_images_to_use:
+                break
+            img_bgr: np.array = cv2.imread(os.path.join(path, filename))
```
Comment (edgarriba) [inline kornia/geometry/nerf/nerf.py:95]: possibly we should start really thinking to have a simple `kornia.io` to leverage image/video reading/writing. Based on recent survey seems that our community is an opencv fan. /cc @shijianjian @ducha-aiki @lferraz 
https://twitter.com/kornia_foss/status/1469232449917968386

File: kornia/geometry/nerf/nerf.py [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+    for p1, p2 in zip(model1.parameters(), model2.parameters()):
+        if p1.data.ne(p2.data).sum() > 0:
+            models_equal = False
+            break
+    return models_equal
+
+import copy
+
+
+class MyTestCase(unittest.TestCase):
```
Comment (edgarriba) [inline kornia/geometry/nerf/nerf.py:275]: for the final version test should go apart and use pytest instead

File: kornia/geometry/nerf/nerfmm/models/intrinsics.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
+import numpy as np
+import torch
+import torch.nn as nn
+
+
+class LearnFocal(nn.Module):
```
Comment (edgarriba) [inline kornia/geometry/nerf/nerfmm/models/intrinsics.py:6]: similar to `PinholeCamera` this could be a class `PinholeCameraIntrinsics` with learnable parameters.

File: kornia/geometry/nerf/nerfmm/models/poses.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
+import torch
+import torch.nn as nn
+
+from kornia.geometry.nerf.nerfmm.utils.lie_group_helper import make_c2w
+
+
+class LearnPose(nn.Module):
```
Comment (edgarriba) [inline kornia/geometry/nerf/nerfmm/models/poses.py:7]: similar to `PinholeCamera` this could be a class `PinholeCameraExtrinsics` or `PinholeCameraPose` with learnable parameters.

File: kornia/geometry/nerf/nerfmm/utils/comp_ray_dir.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
+import torch
+
+
+def comp_ray_dir_cam(H, W, focal):
+    """Compute ray directions in the camera coordinate, which only depends on intrinsics.
+
+    This could be further transformed to world coordinate later, using camera poses.
+    :return: (H, W, 3) torch.float32
+    """
+    y, x = torch.meshgrid(torch.arange(H, dtype=torch.float32),
```
Comment (edgarriba) [inline kornia/geometry/nerf/nerfmm/utils/comp_ray_dir.py:10]: you can use kornia create_meshgrid + convert_to_homogeneous

File: kornia/geometry/nerf/nerfmm/utils/lie_group_helper.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
+import numpy as np
+import torch
+from scipy.spatial.transform import Rotation as RotLib
+
+
+def SO3_to_quat(R):
```
Comment (edgarriba) [inline kornia/geometry/nerf/nerfmm/utils/lie_group_helper.py:6]: check on the conversions module because we have some of this functionalities
https://kornia.readthedocs.io/en/latest/geometry.conversions.html

Comment (edgarriba) [reply kornia/geometry/nerf/nerfmm/utils/lie_group_helper.py:6]: in case of missing we can implement there

File: kornia/geometry/nerf/nerfmm/utils/training_utils.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+def set_randomness(args):
+    if args.true_rand is False:
+        random.seed(args.rand_seed)
+        np.random.seed(args.rand_seed)
+        torch.manual_seed(args.rand_seed)
+        torch.backends.cudnn.deterministic = True
+        torch.backends.cudnn.benchmark = False
+
+
+def mse2psnr(mse):
```
Comment (edgarriba) [inline kornia/geometry/nerf/nerfmm/utils/training_utils.py:26]: we have `ssim_loss`: https://kornia.readthedocs.io/en/latest/losses.html#kornia.losses.ssim_loss

File: kornia/geometry/nerf/nerfmm/utils/volume_op.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+    Outputs:
+        rays_o: (N_rays, 3), the origin of the rays in NDC
+        rays_d: (N_rays, 3), the direction of the rays in NDC
+    """
+    t = -(near + rays_o[..., 2]) / rays_d[..., 2]
+    rays_o = rays_o + t[..., None] * rays_d
+
+    # Store some intermediate homogeneous results
+    ox_oz = rays_o[..., 0] / rays_o[..., 2]
```
Comment (edgarriba) [inline kornia/geometry/nerf/nerfmm/utils/volume_op.py:30]: use convert_points_from_homogenous

File: kornia/geometry/nerf/nerfmm/utils/volume_op.py [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+    :param c2w:             (4, 4)                  camera pose
+    :param ray_dir_cam:     (H, W, 3)               ray directions in the camera coordinate
+    :param t_vals:          (N_sample, )            sample depth in a ray
+    :param perturb_t:       True/False              whether add noise to t
+    """
+    ray_H, ray_W = ray_dir_cam.shape[0], ray_dir_cam.shape[1]
+    N_sam = t_vals.shape[0]
+
+    # transform rays from camera coordinate to world coordinate
+    ray_dir_world = torch.matmul(c2w[:3, :3].view(1, 1, 3, 3),
```
Comment (edgarriba) [inline kornia/geometry/nerf/nerfmm/utils/volume_op.py:104]: use `transform_points` https://kornia.readthedocs.io/en/latest/geometry.linalg.html?highlight=transform_points#kornia.geometry.linalg.transform_points

**Review by edgarriba — COMMENTED** [review]
high-level comments to start the integration with the library. Please, @ducha-aiki @shijianjian @lferraz check it out

File: kornia/geometry/nerf/from_nerfmm.py [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+    rendered_img = torch.cat(rendered_img, dim=0)  # (H, W, 3)
+    rendered_depth = torch.cat(rendered_depth, dim=0)  # (H, W)
+    return rendered_img, rendered_depth
+
+def normalize(v):
+    """Normalize a vector."""
+    return v / np.linalg.norm(v)
+
+
+def create_spiral_poses(radii, focus_depth, n_poses=120, n_circle=2):
```
[the remark this answers is not in the document]
Comment (YanivHollander) [reply kornia/geometry/nerf/from_nerfmm.py:244]: I can open a separate PR for that

File: kornia/geometry/nerf/nerf.py (hunk 1 above)
Comment (YanivHollander) [reply kornia/geometry/nerf/nerf.py:26]: Should be possible to use this DS as a unified container for the estimated camera parameters

File: kornia/geometry/nerf/nerf.py (hunk 3 above)
Comment (YanivHollander) [reply kornia/geometry/nerf/nerf.py:95]: I modified the method `load_images`  to use `torchvision.io.read_image` to load all images in a directory, and then stack them to a 4D tensor. I removed the previous dependency on openCV and Numpy.

This function can be separated and moved to a Kornia module such as `kornia.io`

**Comment by YanivHollander** [comment]
Many of the problems I am seeing with objects other than the flower scene could potentially be related and improved by implementing the idea in the following: [https://jonbarron.info/mipnerf/](https://jonbarron.info/mipnerf/)

**Comment by edgarriba** [comment]
> Many of the problems I am seeing with objects other than the flower scene could potentially be related and improved by implementing the idea in the following: https://jonbarron.info/mipnerf/

seems that the code is in jax though. Good experiment to see the difficulty to port to pytorch

**Comment by bsuleymanov** [comment]
@edgarriba there is a draft version here: https://github.com/bsuleymanov/mip-nerf

**Comment by YanivHollander** [comment]
Hi,

Just to quickly update. I am backlogged by many other tasks that keep me busy, so revising NeRF code takes longer than expected.

I am thinking now that utilizing primitives form PyTorch3D should really be the way to go. It will allow more flexibility moving forward. I am now focusing on this, and trying mainly to solve compilation issues of the package on my Mac. I will update when I have more to share.

If anyone has some experience with PyTotch3D on OS X please share

**Comment by edgarriba** [comment]
> Hi,
> 
> Just to quickly update. I am backlogged by many other tasks that keep me busy, so revising NeRF code takes longer than expected.
> 
> I am thinking now that utilizing primitives form PyTorch3D should really be the way to go. It will allow more flexibility moving forward. I am now focusing on this, and trying mainly to solve compilation issues of the package on my Mac. I will update when I have more to share.
> 
> If anyone has some experience with PyTotch3D on OS X please share

What primitives are we talking about?

**Comment by YanivHollander** [comment]
PyTorch3D package has configurable classes for ray samplings, ray marching, positional embedding, etc. It also has camera models, which we may not need since Kornia has similar DSs. What PyTorch3D doesn't have is the part to estimate camera parameters for the purpose of calibration, so for that part I will keep using Wang's contribution.

More on PyTorch3D NeRF in a tutorial at: [https://pytorch3d.org/tutorials/fit_simple_neural_radiance_field](https://pytorch3d.org/tutorials/fit_simple_neural_radiance_field).

**Comment by edgarriba** [comment]
Not sure if this will go but a recent reference I found: https://towardsdatascience.com/its-nerf-from-nothing-build-a-vanilla-nerf-with-pytorch-7846e4c45666?source=email-5c8c37c62f83-1651368507103-digest.reader-7f60cf5620c9-7846e4c45666----0-98------------------eb0538b6_e4a6_4139_a493_05f89c056444-1-

**Comment by YanivHollander** [comment]
I had to deprioritize my work on NeRF due to many other tasks. On top of that my choice to use `PyTorch3D` may not have been the best, since I suspect there are few bugs in this framework that prevent me from getting results as good as when I was using the (slower, though) Wang, 2011 implementation. I plan to go back to this effort, and will take a look at the reference you sent.

**Comment by ducha-aiki** [comment]
Closing in favour of #1911
